### ISSUE TITLE
glNext improving DeferredShadingAdvanced

### ISSUE THREAD

**Issue by simongeilfus** [cinder/Cinder#783]
Hey all, 
Glad those two samples got merged! Really nice work Steve.

If you don't mind me being a bit more picky, I was wondering if we could do better for the ambient occlusion pass! I had a quick look today and it seems a bit weird to me. Well actually it's a bit hard to see anything with the current geometry as there's really little places for the ao to get obvious. But if you add more overlapping meshes it get pretty clear that the ao pass doesn't add much. From what I can see the SSAO method you are using only creates some black outlines and darker (left) faces, it's probably also related to the parameters of the shader but I would propose to switch to a newer one.

Here's a few comparison between your algorithm and the one I'm using. It's called Scalable Ambient Obscurance and is IMO a huge improvement on most other SSAO methods; it scales pretty well and allow to have a much wider sample radius without performance penalties.

I added a few spheres and cubes to your scene, so we can see a bit more what's going on :

current SSAO:
![frame5140](https://cloud.githubusercontent.com/assets/241098/6963059/e94e128e-d93a-11e4-9f10-13f4f05ab436.png)
SAO:
![frame5611](https://cloud.githubusercontent.com/assets/241098/6963064/ed75110a-d93a-11e4-92f5-695ce2f3f1fe.png)

current SSAO:
![frame5752](https://cloud.githubusercontent.com/assets/241098/6963068/f51a48f8-d93a-11e4-8597-eec47e46f514.png)
SAO:
![frame6074](https://cloud.githubusercontent.com/assets/241098/6963074/fed8656e-d93a-11e4-983c-5ce3aeb523f5.png)

current SSAO:
![frame6827](https://cloud.githubusercontent.com/assets/241098/6963075/032f8822-d93b-11e4-836f-d627cafdf3e2.png)
SAO:
![frame7341](https://cloud.githubusercontent.com/assets/241098/6963078/06e37938-d93b-11e4-813f-cec7eca9cbab.png)

current SSAO:
![frame8616](https://cloud.githubusercontent.com/assets/241098/6963082/12b3a60c-d93b-11e4-8334-b60ac5e6d34d.png)
SAO:
![frame9176](https://cloud.githubusercontent.com/assets/241098/6963083/16636878-d93b-11e4-8705-3c72ef0d2eb5.png)

I'd be happy to send a PR for this but wanted to discuss a few things before. The sample would have to be a bit modified to be able to implement this properly. The above screenshots might look nice, but things are not that easy. The main issue is that this pass is happening at half the resolution. Which totally makes sense. The thing is because your SSAO is working at a really high frequency (small AO radius), blurring the small fbo is enough to blend properly with the rest of the shading, you probably won't see the low-res behind. You basically don't need any upscaling technique. On the other hand as soon as you raise the AO radius, compositing it with the rest will look totally wrong if not upscaled properly. Which means that if we switch to any AO method that is actually visible, we'll probably have to add one more full-res render target and another filtering/upscaling shader.

So in my opinion, we should probably or get rid of the AO pass and make the sample even simpler, or add a GlslProg and a Fbo and make it work nicely. 

Let me know what you think.

Simon.

**Comment by BanTheRewind** [comment]
I'd rather do the latter (have nicer SSAO). Feel free to implement and PR / gist.

**Comment by BanTheRewind** [comment]
I made you a branch:
https://github.com/bantherewind/Cinder/tree/sao

There is a separate, full-screen SAO buffer with its own textures. There is a SAO pass with a mostly empty shader (I left some utilities in there for you). All you have to do is fill in sao.vert/sao.frag. The SAO pass is followed by a blur pass, just like SSAO. If you don't need a blur pass to follow, just delete those passes and I'll reduce the buffer to a single attachment. Otherwise, blur is on for you.  :)

**Comment by richardeakin** [comment]
Should we merge #803 first?

**Comment by BanTheRewind** [comment]
Yes. The sao branch stems from #803. After the new SAO is implemented, there will be one more PR on this where things are more batch-driven and drawn with instancing. Shouldn't really add complexity, but will be a nice performance boost.

**Comment by BanTheRewind** [comment]
Working on this myself, actually... Curious to learn the technique. I see it takes three passes (before blur), so a little more set up is required.

**Comment by BanTheRewind** [comment]
Actually, feel free to implement and PR. Might not be able to get to this this week.

**Comment by BanTheRewind** [comment]
Going to delete the SAO branch for now. It would be hard to merge with the recent updates.

**Comment by simongeilfus** [comment]
Hey Steve!
Sorry busy week here, I'll give it another try this weekend.

**Comment by BanTheRewind** [comment]
Thanks much!




I've made a number of updates that would have made the old SAO branch too tough to merge. Let me know if you need anything. Like if I can set up buffers and passes / batches for you to save some time.

**Comment by BanTheRewind** [comment]
OOooooooOoOoO K
I've implemented HBAO at full res here:
https://github.com/BanTheRewind/Cinder/tree/ao

Needs a little code cleanup, but looks nice. I think SAO will look better! But I wanted to compare for both appearance and performance. @simongeilfus I think the way the project is set up here is a better starting point for you than before. The ambient occlusion pass has its own buffer, batch, etc. Feel free to get in there swap out the AO shaders for your own code.

In the mean time, I'm waiting on an update which will allow me to use mat4 as a per-instance attribute so anything that's multiple may be drawn with instancing (ie, lights, icosahedrons, etc).

 Thanks!

**Comment by BanTheRewind** [comment]
Sorry for making this such a moving target. Everything is just in the glNext branch of my repo now. I've added a moving overhead light which is tied to the shadow camera. There are now four quality modes selectable in the params, and AO blur is optional. I reorked the HBAO shader a ton. Looks really nice and is far more efficient.

I would still like to see a SAO implementation, since HBAO is works with a small sampling radius.

**Comment by BanTheRewind** [comment]
@simongeilfus 
Hey, I've fired this branch back up.
https://github.com/bantherewind/Cinder/tree/sao

I'm working it so you can switch between HBAO and SAO. Might leave them both in there. Might just pick one. Probably SAO.

I know you're busy, but would you mind just taking a peek in there just to confirm I'm on the right path? It looks like I have my mipmap FBO / texture situation working. I'm writing clip space Z from depth into the first level, then using minify shader from McGuire to fill in the mipmap levels. That's as far as I've gotten. I have an issue (that will probably be obvious to resolve when I get home) where the mipmaps are super tiny. Not 100% sure the data is right.

A second set of eyes on the FBO set up and shaders that are there would help.

Thanks much!

- Steve

**Comment by BanTheRewind** [comment]
This is mostly implemented, but am hitting some blocks. Going to think out loud here, but also record this issue in the hope it's seen.  ;)

I set up a new buffer (mFboCsz) which fills in mipmap levels for a texture (mTextureFboCsz). I calculate clip space Z from depth and write it to level 0 (mBatchSaoCszRect). Then I use McGuire's minify shader (adapted for GLSL 330) to fill in mipmap levels 1 - 4 (mBatchSaoMinRect). Then that is used to render the SAO shader into the AO buffer (the one I was using for HBAO) (mBatchSaoAoRect).

I think pretty much everything is right now except my CSZ pass. I even have the guard band and offsets in there. If I just pass through the depth image in the CSZ pass (no clip-space reconstruction), I see it making through the SAO chain onto the screen (including custom mipmap filtering). Using the math from the paper, my CSZ buffer some comes out white. The entire AO pipeline is just one channel. GL_R32F for the mipmapped texture and some other one channel float (depending on selected render quality) for the final buffer where the AO is rendered.

I feel like I may be missing something in either the data type for the CSZ buffer or in the clip and projection info I'm passing in to the CSZ and/or SAO shaders. I'm working from McGuire's sample and things look right.  Hmmm...

Side note. I think it's best to be able to toggle between no AO, HBAO, and SAO. I think the purpose of this sample is to both educate what you can do on top of a deferred shading pipeline, but also borrow from it (ie, rendering technique grab bag). HBAO isn't as pretty or efficient as SAO, but it's *dead simple* to implement. It's the best looking version of the AO methods which only require one buffer and one shader, IMO. SAO is probably the best method, period, but between the mipmap setup and compensating for the guard band, it's not easily reusable for beginners.

I'll stop short of turning this into a tiled deferred renderer.  ;)

**Comment by meshula** [comment]
> I'll stop short of turning this into a tiled deferred renderer. ;)

Why? ;)

**Comment by simongeilfus** [comment]
Hey Stephen, sorry for not be able to commit more time on this, I had a quick look at your mipmap issue and can't see anything wrong,... I'll double check this weekend and let you know if I find anything!

In the meantime here's the quick SAO I did the other day, it should work as it in the current glNext branch without touching too much at the c++ (if at all)... it's not the full sao with all the passes that you are trying to implement but it does work:
https://gist.github.com/simongeilfus/54ac187bada4b420efd5
And here the unfinished bilateral blur (not working yet) : https://gist.github.com/simongeilfus/c42fe5eb1c9586492d64

**Comment by simongeilfus** [comment]
Haha! Cross-posting!

**Comment by BanTheRewind** [comment]
Ha! I meant to say I'll stop short of turning this sample into a tiled deferred shading engine. I absolutely plan to make this tiled for my personal engine.  :)

**Comment by BanTheRewind** [comment]
Thanks!

**Comment by simongeilfus** [comment]
Hey Stephen, could you send me your email (mine is my firstnamelastname@gmail.com)  ... I should really really show you what I'm working on atm...

**Comment by BanTheRewind** [comment]
Oh, awesome on the SAO blur. I'll make sure it uses that when the AO method is SAO.

**Comment by BanTheRewind** [comment]
Ah, I see... Check out the "presentation" link here:
http://graphics.cs.williams.edu/papers/SAOHPG12/

SAO definitely looks better, but it can also perform *way* better than other methods with a little prep work. The paper (same link) goes into detail about an optimization which speeds up the process by a factor of seven. I think my failure is happening in the first pass which stores clip-space Z.

**Comment by BanTheRewind** [comment]
I'm glad I thought out loud here. I totally have it working! The CSZ pass needed some scaling in the clip info. Needs some cleanup and some scaling math, but it works. And it's super fast! At 2560x1440 with a GTX 680M, I get 52 fps with SAO vs 48 with HBAO. 58fps with no AO. That includes blurring.

**Comment by simongeilfus** [comment]
Exiting! Can't wait to check this out.... 

In the meantime, as you seem to be crunching white papers at a really scary speed you might want to check [this out](http://frictionalgames.blogspot.be/2014/01/tech-feature-ssao-and-temporal-blur.html) next. IMO Temporal supersampling is probably the most exiting thing in recent papers. In the case of your SAO implementation, it would allow to do the AO at a much lower res and have it basically for free. (Same idea behind [Epic crazy antialiasing](https://www.youtube.com/watch?v=yNQ47MY-Eo0)),... definitely out of this sample scope but highly recommended to read about this technique. 

(Btw the code I posted above is adapted from McGuire sources)

**Comment by BanTheRewind** [comment]
We can close this. HBAO and SAO are implemented.
